### PR TITLE
Revert tags as metadata

### LINKED ISSUE #5871: Race condition between ManagedMetadataProvider and ItemRegistryImpl wrt tags
It looks like there is a possible race condition between the `ManagedMetadataProvider` providing its content to the `MetadataRegistry` (including tags) and the `ItemRegistryImpl` doing the migration of tags which were previously persisted in the database:

```
2018-07-06T15:11:47.132+0200   147  DEBUG          [symbolicName=****, className=****] - Initializing new custom db with alias org.eclipse.smarthome.core.items.Metadata
2018-07-06T15:11:47.247+0200   147  DEBUG          [symbolicName=****, className=****] - Initializing new custom db with alias org.eclipse.smarthome.core.items.Item
2018-07-06T15:11:47.963+0200   135  WARNING        [symbolicName=org.eclipse.smarthome.core, className=org.eclipse.smarthome.core.common.registry.AbstractRegistry] - Could not add element: Cannot add element, because an element with same UID (_tags:hz_1) already exists.
java.lang.IllegalArgumentException: Cannot add element, because an element with same UID (_tags:hz_1) already exists.
	at org.eclipse.smarthome.core.common.registry.AbstractManagedProvider.add(AbstractManagedProvider.java:56)
	at org.eclipse.smarthome.core.common.registry.AbstractRegistry.add(AbstractRegistry.java:222)
	at org.eclipse.smarthome.core.internal.items.ItemRegistryImpl.writeTags(ItemRegistryImpl.java:465)
	at org.eclipse.smarthome.core.internal.items.ItemRegistryImpl.addTags(ItemRegistryImpl.java:488)
	at org.eclipse.smarthome.core.internal.items.ItemRegistryImpl.onAddElement(ItemRegistryImpl.java:298)
	at org.eclipse.smarthome.core.internal.items.ItemRegistryImpl.onAddElement(ItemRegistryImpl.java:1)
	at org.eclipse.smarthome.core.common.registry.AbstractRegistry.addProvider(AbstractRegistry.java:305)
	at org.eclipse.smarthome.core.internal.items.ItemRegistryImpl.addProvider(ItemRegistryImpl.java:562)
	at org.eclipse.smarthome.core.common.registry.AbstractRegistry$ProviderTracker.addingService(AbstractRegistry.java:114)
	at org.eclipse.smarthome.core.common.registry.AbstractRegistry$ProviderTracker.addingService(AbstractRegistry.java:1)
    ****
[several more of the warnings above]
2018-07-06T15:11:49.346+0200   135  DEBUG          [symbolicName=org.eclipse.smarthome.core, className=org.eclipse.smarthome.core.common.registry.AbstractRegistry] - Provider 'org.eclipse.smarthome.core.items.ManagedItemProvider' has been added.
2018-07-06T15:11:49.464+0200   135  DEBUG          [symbolicName=org.eclipse.smarthome.core, className=org.eclipse.smarthome.core.common.registry.AbstractRegistry] - Provider 'org.eclipse.smarthome.core.internal.items.ManagedMetadataProviderImpl' has been added.
```

### LINKED ISSUE #5872: The "silent" tag migration is conceptually broken
The `ItemRegistryImpl` does a somewhat "silent" migration of tags from items where they were previously persisted into the metadata storage.

However, this is conceptually flawed, as the tags remain persisted together with the item (as they were before #5431), i.e. if tags then are modified (especially removed) in the metadata, they will re-appear again after a restart. 

The following test-case illustrates this behavior:

```java
    @Test
    public void testReloadWithFormerlyPersistedTags() {
        prepareMetadata(ITEM_NAME, "foo");
        prepareItem(ITEM_NAME, "bar");

        // fake an item with previously persisted tag
        Object persistedObject = volatileStorageService.getStorage(Item.class.getName()).get(ITEM_NAME);
        ((PersistedItem) persistedObject).tags = Collections.singleton("bar");

        itemRegistry.removeTag(ITEM_NAME, "bar");
        assertEquals(1, itemRegistry.get(ITEM_NAME).getTags().size());

        // simulate a "reboot" by unregistering & re-registering the volatile storage service
        unregisterService(StorageService.class.getName());
        waitForAssert(() -> assertNull(itemRegistry.get(ITEM_NAME)));
        registerService(volatileStorageService, StorageService.class.getName());
        waitForAssert(() -> assertNotNull(itemRegistry.get(ITEM_NAME)));

        Item item = itemRegistry.get(ITEM_NAME);
        assertEquals(1, item.getTags().size()); // <--- THIS IS WERE IT FAILS
        assertEquals("foo", item.getTags().iterator().next());
    }
```

### PR DESCRIPTION
**Pull request by sjka**
The move of tags from being integral part of the items to re-using the metadata infrastructure was supposed to be purely an internal implementation detail. However, there have been several flaws now with #5871 and #5872 being only the latest ones. #5861 also was a hack to prevent service lifecycle issues by adding a dependency between to services which actually shouldn't know each other. Also, #5852 is not capable of notifying listeners about changes in tags - "before" and "after" are always identical regardless if they were the cause for that change or not. 

Therefore I came to the conclusion that it simply was not a good idea to try making that invisibly to the outside without breaking the API. IMHO it would be better to keep the old, tags-on-item infrastructure as it was. Maybe deprecating it.

Nevertheless, it is open for any solution to actually use the metadata infrastructure to maintain tags in there - but then really separate, as what they are: metadata. With the corresponding contract. With the corresponding expectations wrt lifecycles. 

In any case, I don't really see how the current approach can be fixed to become smooth, elegant and fully working. Squeezing two separate entities into pretending to be one, with the same contract and lifecycle, does not seem possible with reasonable effort. This is not gonna end well!

Therefore, this PR effectively rolls-back the "Item tags become metadata"-PR and all subsequent, insufficient attempts to 

reverts #5431 
reverts #5852 
reverts #5861 
reverts #5495 
reverts #5625 
reverts #5652 
reverts #5704
reverts #5858
fixes #5871 
fixes #5872 
Signed-off-by: Simon Kaufmann <simon.kfm@googlemail.com>

### CODE REVIEW

**Comment by afuechsel** [comment]
+1

**Comment by afuechsel** [comment]
I fully agree with your findings. :) 

Will ESH then eventually move away from using tags on items at all (mark them as deprecated) and only use metadata in the future? I think, using metadata is more flexible than only using tags. How does this relate to the discussion about Semantic Tagging and this: https://github.com/eclipse/smarthome/wiki/Semantic-Tag-Library

**Comment by maggu2810** [comment]
@sjka Okay for me. Have you already discussed with @kaikreuzer 
I am pretty sure that changes shouldn't be merged if he does not agree.
WDYT?

**Comment by sjka** [comment]
Yes, absolutely! 

I did - and I think I could convince him. Our common understanding so far is to keep the current item-bound tags in place and, as a next step, just deprecate them in favor of maintaining tags as metadata directly. 
However, the details are yet to be discussed: whether it should be "just" metadata with some commonly agreed namespace & format, or if (and where) there should be a dedicated API which helps maintaining the tags correctly. Also, @afuechsel's point about the relation to the semantic tagging should be clarified (elsewhere 😉), i.e. if some arbitrary user-provided tags should be mixed with the "semantic" tags or not.

Nevertheless, I would appreciate if you (i.e. @kaikreuzer) would express your consent here yourself, once time is permitting.

**Comment by kaikreuzer** [comment]
~~ _expressing consent_ ~~ 🏳️ 🏳️ 🏳️ 

It would have been nice if the transition could have happened invisible to the consumer and without any deprecation or new APIs. But I fully trust @sjka's analysis that this was probably a too optimistic dream and not feasible in reality. So I am ok with rolling this back and keeping the tags on the items for the time being; seeing it positively, it might be a good chance to deprecate the tags as they are today and replace them by the new pre-defined semantic tagging (as metadata), which will have more requirements wrt querying and filtering etc. anyhow.

Note: I didn't review the PR itself, so this should be done by someone else, please!

File: bundles/model/org.eclipse.smarthome.model.item/src/org/eclipse/smarthome/model/item/internal/GenericMetadataProvider.java [hunk 1]
```diff
@@ -119,7 +86,7 @@ public void removeMetadata(String itemName) {
     public Collection<Metadata> getAll() {
         try {
             lock.readLock().lock();
-            return Collections.unmodifiableCollection(metadata.values());
+            return Collections.unmodifiableSet(metadata);
```
Comment (maggu2810) [inline bundles/model/org.eclipse.smarthome.model.item/src/org/eclipse/smarthome/model/item/internal/GenericMetadataProvider.java]:
```java
return Collections.unmodifiableSet(new HashSet<>(metadata));
```

**Comment by maggu2810** [comment]
@sjka, @kaikreuzer I agree with your comments above, too.

**Pull request merged**